node-backend/scenarioHandler.js

```diff
@@ -82,47 +83,62 @@ module.exports = function(db) {
+    db.runInTransaction(_.partial(deleteTransaction, workspaceId, subproblemId, scenarioId), function(error) {
       if (error) {
         util.handleError(error, next);
       } else {
-        logger.debug('done deleting subProblem : ' + JSON.stringify(subproblemId));
+        logger.debug('Done deleting scenario: ' + JSON.stringify(scenarioId));
         response.sendStatus(httpStatus.OK);
       }
     });
   }
 
-  function deleteTransaction(subproblemId, scenarioId, next, client, transactionCallback) {
+  function deleteTransaction(workspaceId, subproblemId, scenarioId, client, transactionCallback) {
     async.waterfall([
-      _.partial(blockIfOnlyOneScenario, subproblemId, next),
-      _.partial(deleteScenarioAction, scenarioId, next)
+      _.partial(getScenarioIds, subproblemId),
+      _.partial(getDefaultScenario, workspaceId),
+      _.partial(setDefaultScenario, workspaceId, scenarioId),
+      _.partial(deleteScenarioAction, scenarioId)
     ], transactionCallback);
   }
 
-  function blockIfOnlyOneScenario(subproblemId, next, callback) {
-    ScenarioRepository.countScenariosForSubproblem(subproblemId, function(error, result) {
-      if (error) {
-        util.handleError(error, next);
-      } else if (result.rows[0] === 1) {
-        util.handleError('Cannot delete the only scenario for subproblem', next);
-      } else {
-        callback(null);
-      }
-    });
+  function getScenarioIds(subproblemId, callback) {
+    ScenarioRepository.getScenarioIdsForSubproblem(subproblemId, callback);
   }
 
-  function deleteScenarioAction(subproblemId, next, callback) {
-    ScenarioRepository.delete(
-      subproblemId,
-      function(error) {
-        if (error) {
-          util.handleError(error, next);
-        } else {
-          callback(null);
-        }
-      }
-    );
+  function getDefaultScenario(workspaceId, scenarioIds, callback) {
+    if (scenarioIds && scenarioIds.rows.length === 1) {
+      callback('Cannot delete the only scenario for subproblem');
+    } else {
+      WorkspaceRepository.getDefaultScenario(workspaceId, function(error, result) {
+        callback(error, result, scenarioIds.rows);
+      });
+    }
+  }
+
+  function setDefaultScenario(workspaceId, scenarioId, defaultId, scenarioIds, callback) {
+    if (defaultId.rows[0].defaultscenarioid + '' === scenarioId) {
+      const newDefaultScenario = getNewDefaultScenario(scenarioIds, scenarioId);
+      changeDefaultScenarioId(workspaceId, newDefaultScenario, callback);
+    } else {
+      callback(null, null);
+    }
+  }
+  
+  function getNewDefaultScenario(scenarioIds, scenarioId) {
+    return _.find(scenarioIds, function(row) {
+      return row.id + '' !== scenarioId;
+    }).id;
+  }
+
+  function changeDefaultScenarioId(workspaceId, scenarioId, callback) {
+    WorkspaceRepository.setDefaultScenario(workspaceId, scenarioId, callback);
+  }
+
+  function deleteScenarioAction(scenarioId, setDefaultScenarioResult, callback) {
```

Comment:
I'm not sure, but I *think* that if the `setDefaultScenario` doesn't get called back with an actual value you don't need the redundant argument here and can inline the deleteScenarioAction as well.

Reply:
Would need to see what you mean to act on it.